man/man8/xfs_healer.8
.TH xfs_healer 8
.SH NAME
xfs_healer \- automatically heal damage to XFS filesystem metadata
.SH SYNOPSIS
.B xfs_healer
[
.B OPTIONS
]
.I mount-point
.br
.B xfs_healer \-V
.SH DESCRIPTION
.B xfs_healer
is a daemon that tries to automatically repair damaged XFS filesystem metadata.
.PP
.B WARNING!
This program is
.BR EXPERIMENTAL ","
which means that its behavior and interface
could change at any time!
.PP
.B xfs_healer
asks the kernel to report all observations of corrupt metadata, media errors,
filesystem shutdowns, and file I/O errors.
The program can respond to runtime metadata corruption errors by initiating
targeted repairs of the suspect metadata or a full online fsck of the
filesystem.

Normally this program runs as a systemd service.
The service is activated via the
.I xfs_healer_start
service if systemd is supported.

The kernel may not support repairing or optimizing the filesystem.
If this is the case, the filesystem must be unmounted and
.BR xfs_repair (8)
run on the filesystem to fix the problems.
.SH OPTIONS
.TP
.BI \-\-everything
Ask the kernel to send us good metadata health events, not only events related
to metadata corruption, media errors, shutdowns, and I/O errors.
.TP
.B \-\-foreground
Start enough event handling threads to allow consumption of all online CPUs.
If not specified, start exactly one event handling thread.
.TP
.B \-\-no-autofsck
Do not use the
.I autofsck
filesystem property to decide whether or not to repair corrupt metadata.
If the
.B \-\-repair
option is given, then all corruptions will be repaired.
If the
.B \-\-repair
option is not given, then the program will never try to repair the filesystem.
.TP
.B \-\-quiet
Do not print every event to standard output.
.TP
.B \-\-repair
Always try to repair each piece of corrupt metadata when the kernel tells us
about it.
If an individual repair fails or the kernel tells us that health events were
lost, the
.I xfs_scrub
service for this mount point will be launched.
The default is not to try to repair anything.
If this option is specified but the kernel does not support repairs, the
program will exit.
.TP
.B \-\-supported
Check if the filesystem supports sending health events.
Exits with 0 if it does, and non-zero if not.
.TP
.BI \-V
Prints the version number and exit.

.SH AUTOFSCK
By default, this program will read the
.I autofsck
filesystem property to decide if it should try to repair corruptions.
If the property is set to the value
.B repair
then corruptions will be repaired.
If the property is not set but the filesystem supports all back-reference
metadata (reverse mappings and parent pointers), then corruptions will be
repaired.

See the
.BR xfs_scrub (8)
manual page for more details on this filesystem property.

.SH CAVEATS
.B xfs_healer
is an immature utility!
Do not run this program unless you have backups of your data!
This program takes advantage of in-kernel scrubbing to verify a given
data structure with locks held and can keep the filesystem busy for a
long time.
The kernel must be new enough to support the SCRUB_METADATA ioctl.
.PP
If errors are found and cannot be repaired, the filesystem must be
unmounted and repaired.
.SH SEE ALSO
.BR xfs_repair (8)
and
.BR xfs_scrub (8).


debian/changelog
xfsprogs (6.1.0-rc0~WIP-2022-10-14-1djwong1) jammy-experimental; urgency=low

  * Base off for-next on 2022-08-30 (5.19.0+)
  * Support bmbt repairs in xfs_repair.
  * Support realtime rmap btree.
  * Metadata directories.
  * Support atomic file range swaps.
  * AG status and noalloc.
  * Realtime reflink.
  * Better balanced repair workers for scrub.
  * Scrub repair dependencies.
  * reflink/rmapbt/finobt/metadir upgrades.
  * Vectorized scrub experiment.
  * Support for cowextsize on rtreflink files.
  * GETFSREFCOUNTS ioctl and xfs_io utility.
  * Static checking is clean enough (2021-07-15).
  * Prototype free space defragmentation.
  * In-memory reverse mapping btrees for repair.
  * Atomic symlink swapext.
  * Sample mkfs configuration files.
  * Faster xfs_scrub phase5.
  * Scrub file data once per month without a bunch of competing timers.
  * Larger minimum log size and drop support for small filesystems.
  * Multiarch support for libhandle.
  * Better security for xfs_scrub background services.
  * Improve xfs_scrub_all resiliency when systemd gets re-exec'd.
  * Realtime allocation groups.
  * Fixed CPU usage resource control for xfs_scrub processes.
  * Intent items take perag/rtgroup references.

 -- Darrick J. Wong <djwong@djwong.org>  Fri, 14 Oct 2022 09:58:00 -0700

xfsprogs (5.19.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 12 Aug 2022 13:40:25 -0500

xfsprogs (5.18.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 03 Jun 2022 15:18:30 -0400

xfsprogs (5.16.0-1) unstable; urgency=medium

  * New upstream release

  [ Bastian Germann ]
  * debian: Generate .gitcensus instead of .census (Closes: #999743)

 -- Nathan Scott <nathans@debian.org>  Wed, 04 May 2022 14:50:48 -0400

xfsprogs (5.15.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Wed, 06 Apr 2022 16:31:00 -0400

xfsprogs (5.14.2-1) unstable; urgency=medium

  * New upstream release
  * Move rogue fallthrough macro out of linux.h (Closes: #1000974)

 -- Nathan Scott <nathans@debian.org>  Mon, 06 Dec 2021 14:26:57 -0500

xfsprogs (5.14.1-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Thu, 02 Dec 2021 15:26:38 -0500

xfsprogs (5.14.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 19 Nov 2021 10:51:02 -0500

xfsprogs (5.14.0-rc1-1) unstable; urgency=medium

  [ Dave Chinner ]
  * New build dependency: liburcu-dev

  [ Helmut Grohne ]
  * Fix FTCBFS (Closes: #794158)
    + Pass --build and --host to configure

  [ Boian Bonev ]
  * Fix FTBFS (Closes: #997656)
    + Keep custom install-sh after autogen

 -- Bastian Germann <bage@debian.org>  Sun, 14 Nov 2021 23:18:22 +0100

xfsprogs (5.13.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 20 Aug 2021 12:03:31 -0400

xfsprogs (5.12.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 21 May 2021 15:52:00 -0400

xfsprogs (5.11.0-1) unstable; urgency=medium

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 12 Mar 2021 14:08:29 -0500

xfsprogs (5.11.0-rc0-1) experimental; urgency=medium

  [ Steve Langasek ]
  * Regenerate config.guess using debhelper

  [ Bastian Germann ]
  * Build-depend on libinih-dev with udeb package

 -- Bastian Germann <bastiangermann@fishpost.de>  Sat, 20 Feb 2021 11:57:31 +0100

xfsprogs (5.10.0-3) unstable; urgency=medium

  * Drop unused dh-python from Build-Depends (Closes: #981361)
  * Only build for Linux
  * Prevent installing duplicate changelog (Closes: #570704)

 -- Bastian Germann <bastiangermann@fishpost.de>  Fri, 05 Feb 2021 00:18:31 +0100

xfsprogs (5.10.0-2) unstable; urgency=low

  * Team upload
  * debian: cryptographically verify upstream tarball (Closes: #979644)
  * debian: remove dependency on essential util-linux
  * debian: remove "Priority: extra"
  * debian: use Package-Type over its predecessor
  * debian: add missing copyright info (Closes: #979653)

 -- Bastian Germann <bastiangermann@fishpost.de>  Thu, 14 Jan 2021 18:59:14 +0100

xfsprogs (5.10.0-1) unstable; urgency=low

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 11 Dec 2020 17:16:52 -0500

xfsprogs (5.10.0-rc1-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Fri, 04 Dec 2020 12:19:10 -0500

xfsprogs (5.10.0-rc0-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Mon, 16 Nov 2020 14:26:14 -0600

xfsprogs (5.9.0-1) unstable; urgency=low

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Tue, 20 Oct 2020 11:38:49 -0400

xfsprogs (5.9.0-rc1-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Tue, 13 Oct 2020 12:18:43 -0400

xfsprogs (5.9.0-rc0-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Thu, 17 Sep 2020 09:11:13 -0500

xfsprogs (5.8.0-1) unstable; urgency=low

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 04 Sep 2020 15:40:01 -0400

xfsprogs (5.8.0-rc1-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Wed, 26 Aug 2020 14:06:48 -0500

xfsprogs (5.8.0-rc0-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Mon, 10 Aug 2020 14:42:12 -0500

xfsprogs (5.7.0-1) unstable; urgency=low

  * New upstream release

 -- Nathan Scott <nathans@debian.org>  Fri, 24 Jul 2020 11:30:00 -0700

xfsprogs (5.7.0-rc1-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Wed, 15 Jul 2020 13:05:24 -0700

xfsprogs (5.7.0-rc0-1) unstable; urgency=low

  * New upstream prerelease

 -- Nathan Scott <nathans@debian.org>  Thu, 07 May 2020 13:06:26 -0500


healer/xfs_healer.c
// SPDX-License-Identifier: GPL-2.0-or-later
/*
 * Copyright (C) 2025-2026 Oracle.  All Rights Reserved.
 * Author: Darrick J. Wong <djwong@kernel.org>
 */
#include "xfs.h"
#include <pthread.h>
#include <stdlib.h>
#include <sys/xattr.h>

#include "platform_defs.h"
#include "libfrog/fsgeom.h"
#include "libfrog/paths.h"
#include "libfrog/healthevent.h"
#include "libfrog/workqueue.h"
#include "libfrog/systemd.h"
#include "libfrog/fsproperties.h"
#include "xfs_healer.h"

/* Program name; needed for libfrog error reports. */
char				*progname = "xfs_healer";

/* Return a health monitoring fd. */
static int
open_health_monitor(
	struct healer_ctx		*ctx,
	int				mnt_fd)
{
	struct xfs_health_monitor	hmo = {
		.format			= XFS_HEALTH_MONITOR_FMT_V0,
	};

	if (ctx->everything)
		hmo.flags |= XFS_HEALTH_MONITOR_VERBOSE;

	return ioctl(mnt_fd, XFS_IOC_HEALTH_MONITOR, &hmo);
}

/* Report either the file handle or its path, if we can. */
void
lookup_path(
	struct healer_ctx			*ctx,
	const struct xfs_health_monitor_event	*hme,
	struct hme_prefix			*pfx)
{
	uint64_t				ino = 0;
	uint32_t				gen = 0;
	int					ret;

	if (!healer_has_parent(ctx))
		return;

	switch (hme->domain) {
	case XFS_HEALTH_MONITOR_DOMAIN_INODE:
		ino = hme->e.inode.ino;
		gen = hme->e.inode.gen;
		break;
	case XFS_HEALTH_MONITOR_DOMAIN_FILERANGE:
		ino = hme->e.filerange.ino;
		gen = hme->e.filerange.gen;
		break;
	default:
		return;
	}

	ret = weakhandle_getpath_for(ctx->wh, ino, gen, pfx->path,
			sizeof(pfx->path));
	if (ret)
		hme_prefix_clear_path(pfx);
}

/* Decide if this event can only be reported upon, and not acted upon. */
static bool
event_not_actionable(
	const struct xfs_health_monitor_event	*hme)
{
	switch (hme->type) {
	case XFS_HEALTH_MONITOR_TYPE_LOST:
	case XFS_HEALTH_MONITOR_TYPE_RUNNING:
	case XFS_HEALTH_MONITOR_TYPE_UNMOUNT:
	case XFS_HEALTH_MONITOR_TYPE_SHUTDOWN:
		return true;
	}

	return false;
}

/* Should this event be logged? */
static bool
event_loggable(
	const struct healer_ctx			*ctx,
	const struct xfs_health_monitor_event	*hme)
{
	return ctx->log || event_not_actionable(hme);
}

/* Are we going to try a repair? */
static inline bool
event_repairable(
	const struct healer_ctx			*ctx,
	const struct xfs_health_monitor_event	*hme)
{
	if (event_not_actionable(hme))
		return false;

	return ctx->want_repair && hme->type == XFS_HEALTH_MONITOR_TYPE_SICK;
}

/* Handle an event asynchronously. */
static void
handle_event(
	struct workqueue		*wq,
	uint32_t			index,
	void				*arg)
{
	struct hme_prefix		pfx;
	struct xfs_health_monitor_event	*hme = arg;
	struct healer_ctx		*ctx = wq->wq_ctx;
	const bool loggable = event_loggable(ctx, hme);
	const bool will_repair = event_repairable(ctx, hme);

	hme_prefix_init(&pfx, ctx->mntpoint);

	/*
	 * Try to look up the file name for the file we're about to log or
	 * about to repair (which always logs).
	 */
	if (loggable || will_repair)
		lookup_path(ctx, hme, &pfx);

	/*
	 * Non-actionable events should always be logged, because they are 100%
	 * informational.
	 */
	if (loggable) {
		pthread_mutex_lock(&ctx->conlock);
		hme_report_event(&pfx, hme);
		pthread_mutex_unlock(&ctx->conlock);
	}

	/*
	 * If we didn't ask for all the metadata reports (including the healthy
	 * ones) and the kernel tells us it lost something, run the full scan.
	 */
	if (hme->type == XFS_HEALTH_MONITOR_TYPE_LOST && !ctx->everything)
		run_full_repair(ctx);

	/* Initiate a repair if appropriate. */
	if (will_repair)
		repair_metadata(ctx, &pfx, hme);

	free(hme);
}

static unsigned int
healer_nproc(
	const struct healer_ctx	*ctx)
{
	/*
	 * By default, use one event handler thread.  In foreground mode,
	 * create one thread per cpu.
	 */
	return ctx->foreground ? platform_nproc() : 1;
}

enum want_repair {
	WR_REPAIR,
	WR_LOG_ONLY,
	WR_EXIT,
};

/* Determine want_repair from the autofsck filesystem property. */
static enum want_repair
want_repair_from_autofsck(
	struct healer_ctx	*ctx)
{
	char			valuebuf[FSPROP_MAX_VALUELEN + 1] = { 0 };
	enum fsprop_autofsck	shval;
	ssize_t			ret;

	/*
	 * Any OS error (including ENODATA) or string parsing error is treated
	 * the same as an unrecognized value.
	 */
	ret = fgetxattr(ctx->mnt.fd, VFS_FSPROP_AUTOFSCK_NAME, valuebuf,
			FSPROP_MAX_VALUELEN);
	if (ret < 0)
		goto no_advice;

	shval = fsprop_autofsck_read(valuebuf);
	switch (shval) {
	case FSPROP_AUTOFSCK_NONE:
		/* don't run at all */
		ret = WR_EXIT;
		break;
	case FSPROP_AUTOFSCK_CHECK:
	case FSPROP_AUTOFSCK_OPTIMIZE:
		/* log events, do not repair */
		ret = WR_LOG_ONLY;
		break;
	case FSPROP_AUTOFSCK_REPAIR:
		/* repair stuff */
		ret = WR_REPAIR;
		break;
	case FSPROP_AUTOFSCK_UNSET:
		goto no_advice;
	}

	return ret;

no_advice:
	/*
	 * For an unrecognized value, log but do not fix runtime corruption if
	 * backref metadata are enabled.  If no backref metadata are available,
	 * the fs is too old so don't run at all.
	 */
	if (healer_has_rmapbt(ctx) || healer_has_parent(ctx))
		return WR_LOG_ONLY;

	return WR_EXIT;
}

enum mon_state {
	MON_START,
	MON_EXIT,
	MON_ERROR,
};

/* Set ourselves up to monitor the given mountpoint for health events. */
static enum mon_state
setup_monitor(
	struct healer_ctx	*ctx)
{
	const long		BUF_SIZE = sysconf(_SC_PAGE_SIZE) * 2;
	int			mon_fd;
	int			ret;

	ret = xfd_open(&ctx->mnt, ctx->mntpoint, O_RDONLY);
	if (ret) {
		perror(ctx->mntpoint);
		return MON_ERROR;
	}

	if (ctx->autofsck) {
		switch (want_repair_from_autofsck(ctx)) {
		case WR_EXIT:
			printf("%s: %s\n", ctx->mntpoint,
 _("Disabling daemon per autofsck directive."));
			fflush(stdout);
			close(ctx->mnt.fd);
			return MON_EXIT;
		case WR_REPAIR:
			ctx->want_repair = 1;
			printf("%s: %s\n", ctx->mntpoint,
 _("Automatically repairing per autofsck directive."));
			fflush(stdout);
			break;
		case WR_LOG_ONLY:
			ctx->want_repair = 0;
			ctx->log = 1;
			printf("%s: %s\n", ctx->mntpoint,
 _("Only logging errors per autofsck directive."));
			fflush(stdout);
			break;
		}
	}

	/* Check that the kernel supports repairs at all. */
	if (ctx->want_repair && !healer_can_repair(ctx)) {
		if (!ctx->autofsck) {
			fprintf(stderr, "%s: %s\n", ctx->mntpoint,
 _("XFS online repair is not supported, exiting"));
			close(ctx->mnt.fd);
			return MON_ERROR;
		}

		printf("%s: %s\n", ctx->mntpoint,
 _("XFS online repair is not supported, will report only"));
		fflush(stdout);
		ctx->want_repair = 0;
	}

	if (ctx->want_repair) {
		/* Check for backref metadata that makes repair effective. */
		if (!healer_has_rmapbt(ctx))
			fprintf(stderr, "%s: %s\n", ctx->mntpoint,
 _("XFS online repair is less effective without rmap btrees."));

		if (!healer_has_parent(ctx))
			fprintf(stderr, "%s: %s\n", ctx->mntpoint,
 _("XFS online repair is less effective without parent pointers."));

	}

	/*
	 * Open weak-referenced file handle to mountpoint so that we can
	 * reconnect to the mountpoint to start repairs or to look up file
	 * paths for logging.
	 */
	if (ctx->want_repair || healer_has_parent(ctx)) {
		ret = weakhandle_alloc(ctx->mnt.fd, ctx->mntpoint,
				ctx->fsname, &ctx->wh);
		if (ret) {
			fprintf(stderr, "%s: %s: %s\n", ctx->mntpoint,
					_("creating weak fshandle"),
					strerror(errno));
			return MON_ERROR;
		}
	}

	/*
	 * Open the health monitor, then close the mountpoint to avoid pinning
	 * it.  We can reconnect later if need be.
	 */
	mon_fd = open_health_monitor(ctx, ctx->mnt.fd);
	close(ctx->mnt.fd);
	ctx->mnt.fd = -1;
	if (mon_fd < 0) {
		switch (errno) {
		case ENOTTY:
		case EOPNOTSUPP:
			fprintf(stderr, "%s: %s\n", ctx->mntpoint,
 _("XFS health monitoring not supported."));
			break;
		case EEXIST:
			fprintf(stderr, "%s: %s\n", ctx->mntpoint,
 _("XFS health monitoring already running."));
			break;
		default:
			perror(ctx->mntpoint);
			break;
		}
		return MON_ERROR;
	}

	/*
	 * At this point, we know that the kernel is capable of repairing the
	 * filesystem and telling us that it needs repairs.  If the user only
	 * wanted us to check for the capability, we're done.
	 */
	if (ctx->support_check) {
		close(mon_fd);
		return MON_EXIT;
	}

	/*
	 * mon_fp consumes mon_fd.  We intentionally leave mon_fp attached to
	 * the context so that we keep the monitoring fd open until we've torn
	 * down all the background threads.
	 */
	ctx->mon_fp = fdopen(mon_fd, "r");
	if (!ctx->mon_fp) {
		close(mon_fd);
		perror(ctx->mntpoint);
		return MON_ERROR;
	}

	/* Increase the buffer size so that we can reduce kernel calls */
	ctx->mon_buf = malloc(BUF_SIZE);
	if (ctx->mon_buf)
		setvbuf(ctx->mon_fp, ctx->mon_buf, _IOFBF, BUF_SIZE);

	/*
	 * Queue up to 1MB of events before we stop trying to read events from
	 * the kernel as quickly as we can.  Note that the kernel won't accrue
	 * more than 32K of internal events before it starts dropping them.
	 */
	ret = workqueue_create_bound(&ctx->event_queue, ctx, healer_nproc(ctx),
			1048576 / sizeof(struct xfs_health_monitor_event));
	if (ret) {
		errno = ret;
		fprintf(stderr, "%s: %s: %s\n", ctx->mntpoint,
				_("worker threadpool setup"), strerror(errno));
		return MON_ERROR;
	}
	ctx->queue_active = true;

	return MON_START;
}

/* Monitor the given mountpoint for health events. */
static void
monitor(
	struct healer_ctx	*ctx)
{
	bool			mounted = true;
	size_t			nr;

	do {
		struct xfs_health_monitor_event	*hme;
		int		ret;

		hme = malloc(sizeof(*hme));
		if (!hme) {
			pthread_mutex_lock(&ctx->conlock);
			fprintf(stderr, "%s: %s\n", ctx->mntpoint,
					_("could not allocate event object"));
			pthread_mutex_unlock(&ctx->conlock);
			break;
		}

		nr = fread(hme, sizeof(*hme), 1, ctx->mon_fp);
		if (nr == 0) {
			free(hme);
			break;
		}

		if (hme->type == XFS_HEALTH_MONITOR_TYPE_UNMOUNT)
			mounted = false;

		/* handle_event owns hme if the workqueue_add succeeds */
		ret = workqueue_add(&ctx->event_queue, handle_event, 0, hme);
		if (ret) {
			pthread_mutex_lock(&ctx->conlock);
			fprintf(stderr, "%s: %s: %s\n", ctx->mntpoint,
					_("could not queue event object"),
					strerror(ret));
			pthread_mutex_unlock(&ctx->conlock);
			free(hme);
			break;
		}
	} while (nr > 0 && mounted);
}

/* Tear down all the resources that we created for monitoring */
static void
teardown_monitor(
	struct healer_ctx	*ctx)
{
	if (ctx->queue_active) {
		workqueue_terminate(&ctx->event_queue);
		workqueue_destroy(&ctx->event_queue);
	}
	if (ctx->mon_fp) {
		fclose(ctx->mon_fp);
		ctx->mon_fp = NULL;
	}
	free(ctx->mon_buf);
	weakhandle_free(&ctx->wh);
	ctx->mon_buf = NULL;
}

/*
 * Find the filesystem source name for the mount that we're monitoring.  We
 * don't use the fs_table_ helpers because we might be running in a restricted
 * environment where we cannot access device files at all.
 */
static char *
find_fsname(
	const char	*mntpoint)
{
	struct mntent	*mnt;
	FILE		*mtp;
	char		*ret = NULL;
	char		rpath[PATH_MAX], rmnt_dir[PATH_MAX];

	if (!realpath(mntpoint, rpath))
		return NULL;

	mtp = setmntent(_PATH_PROC_MOUNTS, "r");
	if (mtp == NULL)
		return NULL;

	while ((mnt = getmntent(mtp)) != NULL) {
		if (strcmp(mnt->mnt_type, "xfs"))
			continue;
		if (!realpath(mnt->mnt_dir, rmnt_dir))
			continue;

		if (!strcmp(rpath, rmnt_dir)) {
			ret = strdup(mnt->mnt_fsname);
			break;
		}
	}

	endmntent(mtp);
	return ret;
}

static void __attribute__((noreturn))
usage(void)
{
	fprintf(stderr, "%s %s %s\n", _("Usage:"), progname,
			_("[OPTIONS] mountpoint"));
	fprintf(stderr, "\n");
	fprintf(stderr, _("Options:\n"));
	fprintf(stderr, _("  --debug       Enable debugging messages.\n"));
	fprintf(stderr, _("  --everything  Capture all events.\n"));
	fprintf(stderr, _("  --foreground  Process events as soon as possible.\n"));
	fprintf(stderr, _("  --no-autofsck Do not use the \"autofsck\" fs property to decide to repair.\n"));
	fprintf(stderr, _("  --quiet       Do not log health events to stdout.\n"));
	fprintf(stderr, _("  --repair      Always repair corrupt metadata.\n"));
	fprintf(stderr, _("  --supported   Check that health monitoring is supported.\n"));
	fprintf(stderr, _("  -V            Print version.\n"));

	exit(EXIT_FAILURE);
}

enum long_opt_nr {
	LOPT_DEBUG,
	LOPT_EVERYTHING,
	LOPT_FOREGROUND,
	LOPT_HELP,
	LOPT_NO_AUTOFSCK,
	LOPT_QUIET,
	LOPT_REPAIR,
	LOPT_SUPPORTED,

	LOPT_MAX,
};

int
main(
	int			argc,
	char			**argv)
{
	struct healer_ctx	ctx = {
		.conlock	= (pthread_mutex_t)PTHREAD_MUTEX_INITIALIZER,
		.log		= 1,
		.autofsck	= 1,
	};
	int			option_index;
	int			vflag = 0;
	int			c;
	int			ret;

	progname = basename(argv[0]);
	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);

	struct option long_options[] = {
		[LOPT_DEBUG]	   = {"debug", no_argument, &ctx.debug, 1 },
		[LOPT_EVERYTHING]  = {"everything", no_argument, &ctx.everything, 1 },
		[LOPT_FOREGROUND]  = {"foreground", no_argument, &ctx.foreground, 1 },
		[LOPT_HELP]	   = {"help", no_argument, NULL, 0 },
		[LOPT_NO_AUTOFSCK] = {"no-autofsck", no_argument, &ctx.autofsck, 0 },
		[LOPT_QUIET]	   = {"quiet", no_argument, &ctx.log, 0 },
		[LOPT_REPAIR]	   = {"repair", no_argument, &ctx.want_repair, 1 },
		[LOPT_SUPPORTED]   = {"supported", no_argument, &ctx.support_check, 1 },

		[LOPT_MAX]	   = {NULL, 0, NULL, 0 },
	};

	while ((c = getopt_long(argc, argv, "V", long_options, &option_index))
			!= EOF) {
		switch (c) {
		case 0:
			switch (option_index) {
			case LOPT_HELP:
				usage();
				break;
			default:
				break;
			}
			break;
		case 'V':
			vflag++;
			break;
		default:
			usage();
			break;
		}
	}

	if (vflag) {
		fprintf(stdout, "%s %s %s\n", progname, _("version"), VERSION);
		fflush(stdout);
		return EXIT_SUCCESS;
	}

	if (optind != argc - 1)
		usage();
	if (ctx.want_repair)
		ctx.autofsck = 0;

	ctx.mntpoint = argv[optind];
	ctx.fsname = find_fsname(ctx.mntpoint);
	if (!ctx.fsname) {
		fprintf(stderr, "%s: %s\n", ctx.mntpoint,
				_("Not a XFS mount point."));
		ret = -1;
		goto out;
	}

	switch (setup_monitor(&ctx)) {
	case MON_ERROR:
		ret = -1;
		break;
	case MON_EXIT:
		ret = 0;
		break;
	case MON_START:
		ret = 0;
		monitor(&ctx);
		break;
	}

	teardown_monitor(&ctx);
	free((char *)ctx.fsname);
out:
	if (ctx.support_check)
		return systemd_service_exit_now(ret);
	return systemd_service_exit(ret);
}
